container/docker/handler.go

```diff
@@ -218,6 +222,7 @@ func newDockerContainerHandler(
 		Aliases:   []string{strings.TrimPrefix(ctnr.Name, "/"), id},
 		Namespace: DockerNamespace,
 	}
+	handler.ulimits = ctnr.HostConfig.Ulimits
```

Comment:
What would we need to do to get Ulimits in a container-runtime agnostic way?

Reply:
As mentioned in the PR: An alternative approach that might be more "container agnostic" involves looking up all the processes in the cgroup, and emitting the ulimits for the "parent/init/session" process for the cgroup. However, it might be more "fragile".

Also, I'm not sure if container ulimit should be inferred for container-runtimes that don't explicitly provide the ability to set ulimits.

Reply:
Right.  Why would it be more fragile?  Are you suggesting a change in container runtime implementation would break the metric?

How does the process -> ulimit lookup work?

Reply:
The difference is that in this case we're reading the value from the container config. It's what the owner configured at container launch time.

For process->ulimit we'd have find the processes in the cgroups by looking at /sys/fs/cgroup/*/<container>/cgroup.procs, then selecting the process and reading the limits from /proc/<pid>/limits

To do it correctly we probably need to read all the cgroups and find the lowest number pid across all of them. The lowest pid is not necessarily the first pid since pids can wrap around, so ideally we would use some equivalent of pstree, or get the parent pid from the container runtime (I haven't looked into how to do this part).

Reply:
I did a little bit of digging, and I think it should be easy to implement this in a container-runtime agnostic way.

We already get the PID of the container for all container runtimes.  For docker, it is [ctnr.State.Pid](https://github.com/google/cadvisor/blob/master/container/docker/handler.go#L212).

I checked to make sure the PID has the limits specified with `--ulimit` applied to it, following the [docker run example in the docs](https://docs.docker.com/engine/reference/commandline/run/#set-ulimits-in-container---ulimit):

$ docker run --ulimit nofile=1024:1024 --rm debian sh -c "sleep 60; ulimit -n" &
$ docker ps | grep debian 
42048ce156e3 .....
$ docker inspect 42048ce156e3 | grep Pid
            "Pid": 2834013,
            "PidMode": "",
            "PidsLimit": 0,
$ cat /proc/2834013/limits | grep "open files"
Max open files            1024                 1024                 files

I did the same for a container without --ulimit specified, and got:
Max open files            1048576              1048576              files

That shows that the limits are being applied to the PID specified in container.State.Pid, which is what we have access to in cAdvisor already.

To implement this, we would want to add this logic to https://github.com/google/cadvisor/blob/master/container/libcontainer/handler.go#L122.

Reply:
Force pushed a container agnostic version per your suggestions.

Reply:
@dashpole : ping. updated to be container agnostic per your suggestion

Reply:
Thanks.